api/src/services/Database.ts

```diff
@@ -24,126 +20,168 @@ interface Pid {
     pid: number;
 }
 
-async function createDatabase(db): Promise<void> {
-    console.log("Database:createTables");
-    await db.schema.dropTableIfExists("users");
-    await db.schema.createTable("users", (table) => {
-        table.increments("id");
-        table.string("username");
-        table.string("password");
-        table.string("hash");
-    });
-    await db.schema.dropTableIfExists("tokens");
-    await db.schema.createTable("tokens", (table) => {
-        table.string("token").primary();
-        table.timestamp("created_at").notNullable();
-        table.integer("user_id").unsigned().references("users.id");
-    });
-    await db.schema.dropTableIfExists("pids");
-    await db.schema.createTable("pids", (table) => {
-        table.string("namespace").primary();
-        table.integer("pid");
-    });
-    const config = Config.getInstance();
-    const initialPid: Pid = {
-        namespace: config.fedoraPidNameSpace,
-        pid: config.initialPidValue,
-    };
-    await db("pids").insert(initialPid);
-
-    const users = [
-        { username: "geoff", password: "earth", hash: "CuhFfwkebs3RKr1Zo_Do_" },
-        { username: "chris", password: "air", hash: "V1StGXR8_Z5jdHi6B-myT" },
-        { username: "dkatz", password: "avatar", hash: "_HPZZ6uCouEU5jy-AYrDd" },
-    ];
-    for (const user of users) {
-        console.log(`Database:insert: ${user.username}`);
-        await db("users").insert(user);
+class Database {
+    private static instance: Database;
+    config: Config;
+    connection: Knex = null;
+    tokenLifetime = 24 * 60 * 60 * 1000;
+
+    constructor(config: Config) {
+        this.config = config;
+    }
+
+    public static getInstance(): Database {
+        if (!Database.instance) {
+            Database.instance = new Database(Config.getInstance());
+        }
+        return Database.instance;
     }
-}
 
-let db: Knex = null;
-const REBUILD_DB = false;
-async function getDatabase(): Promise<Knex> {
-    if (db === null) {
+    protected async connect(): Promise<void> {
         console.log("Database:connect");
 
         const config: Knex.Config = {
-            client: "sqlite3",
-            connection: {
-                filename: dbFilename,
-            },
+            client: this.config.databaseClient,
+            connection: this.config.databaseConnectionSettings,
             useNullAsDefault: true,
         };
 
-        db = await knex(config);
+        this.connection = await knex(config);
 
-        if (REBUILD_DB || !fs.existsSync(dbFilename)) {
+        await this.initialize(this.connection);
+
+        console.log("Database:ready");
+    }
+
+    protected async initialize(db): Promise<void> {
+        // Special case: if we're using a disk-based database, make sure the containing directory exists:
+        if (this.config.databaseClient === "sqlite3") {
```

Comment:
Is there a reason to keep sqlite3 if we're keeping mysql?

Reply:
Since we're going to be open-sourcing this tool, I think it's useful to offer multiple options... and sqlite is a good tool for evaluating the software, since it requires less setup to get up and running quickly. That being said, I don't think it's critically important, but I figure it doesn't hurt to keep it working if it's not too expensive to do so.